mpc-core/src/protocols/rep3_ring/casts.rs

```diff
@@ -0,0 +1,209 @@
+//! Casts
+//!
+//! Implements casts for sharings of different datatypes
+
+use super::{
+    arithmetic::types::Rep3RingShare,
+    conversion,
+    ring::{bit::Bit, int_ring::IntRing2k},
+    yao,
+};
+use crate::protocols::{
+    rep3::{
+        self,
+        conversion::A2BType,
+        network::{IoContext, Rep3Network},
+        Rep3BigUintShare, Rep3PrimeFieldShare,
+    },
+    rep3_ring::ring::ring_impl::RingElement,
+};
+use ark_ff::PrimeField;
+use num_bigint::BigUint;
+use num_traits::AsPrimitive;
+use rand::{distributions::Standard, prelude::Distribution};
+use std::any::TypeId;
+
+/// Depending on the `A2BType` of the io_context, this function selects the appropriate implementation for the ring cast.
+pub fn ring_cast_selector<T, U, N>(
+    x: Rep3RingShare<T>,
+    io_context: &mut IoContext<N>,
+) -> std::io::Result<Rep3RingShare<U>>
+where
+    T: IntRing2k + AsPrimitive<U>,
+    U: IntRing2k,
+    N: Rep3Network,
+    Standard: Distribution<T> + Distribution<U>,
+{
+    match io_context.a2b_type {
+        A2BType::Direct => cast_a2b(x, io_context),
+        A2BType::Yao => cast_gc(x, io_context),
+        A2BType::StreamingYao => {
```

Comment:
should we just rip out streaming yao?

Reply:
have done so, but kept the backend implementaiton